File: sudoku_generator.js
#!/usr/bin/env node
// sudoku_generator.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');

class Sudoku {
    constructor(grid = null) {
        this.size = 9;
        this.boxSize = 3;
        this.grid = grid ? grid.map(row => [...row]) : Array.from({ length: 9 }, () => Array(9).fill(0));
    }

    copy() {
        return new Sudoku(this.grid.map(row => [...row]));
    }

    isValid(row, col, num) {
        for (let i = 0; i < 9; i++) {
            if (this.grid[row][i] === num || this.grid[i][col] === num) return false;
        }
        const startRow = Math.floor(row / 3) * 3;
        const startCol = Math.floor(col / 3) * 3;
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                if (this.grid[startRow + i][startCol + j] === num) return false;
            }
        }
        return true;
    }

    solve() {
        for (let row = 0; row < 9; row++) {
            for (let col = 0; col < 9; col++) {
                if (this.grid[row][col] === 0) {
                    for (let num = 1; num <= 9; num++) {
                        if (this.isValid(row, col, num)) {
                            this.grid[row][col] = num;
                            if (this.solve()) return true;
                            this.grid[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    countSolutions(limit = 2) {
        let count = 0;
        for (let row = 0; row < 9; row++) {
            for (let col = 0; col < 9; col++) {
                if (this.grid[row][col] === 0) {
                    for (let num = 1; num <= 9; num++) {
                        if (this.isValid(row, col, num)) {
                            this.grid[row][col] = num;
                            count += this.countSolutions(limit);
                            this.grid[row][col] = 0;
                            if (count >= limit) return count;
                        }
                    }
                    return count;
                }
            }
        }
        return 1;
    }

    generate(difficulty = 'medium') {
        this._fillDiagonalBlocks();
        this.solve();
        const cellsToRemove = { easy: 30, medium: 40, hard: 50, expert: 55 }[difficulty] || 40;
        const solution = this.grid.map(row => [...row]);
        const positions = [];
        for (let r = 0; r < 9; r++) {
            for (let c = 0; c < 9; c++) {
                positions.push([r, c]);
            }
        }
        positions.sort(() => Math.random() - 0.5);
        let removed = 0;
        for (const [r, c] of positions) {
            if (removed >= cellsToRemove) break;
            const backup = this.grid[r][c];
            this.grid[r][c] = 0;
            const copy = this.copy();
            if (copy.countSolutions(2) === 1) {
                removed++;
            } else {
                this.grid[r][c] = backup;
            }
        }
        return solution;
    }

    _fillDiagonalBlocks() {
        for (let block = 0; block < 9; block += 3) {
            const nums = [];
            for (let i = 1; i <= 9; i++) nums.push(i);
            for (let i = nums.length - 1; i > 0; i--) {
                const j = Math.floor(Math.random() * (i + 1));
                [nums[i], nums[j]] = [nums[j], nums[i]];
            }
            let idx = 0;
            for (let i = block; i < block + 3; i++) {
                for (let j = block; j < block + 3; j++) {
                    this.grid[i][j] = nums[idx++];
                }
            }
        }
    }

    print(color = true) {
        const reset = color ? '\x1b[0m' : '';
        const blue = color ? '\x1b[34m' : '';
        const white = color ? '\x1b[37m' : '';
        for (let i = 0; i < 9; i++) {
            let line = '';
            for (let j = 0; j < 9; j++) {
                const val = this.grid[i][j];
                const ch = val === 0 ? '.' : String(val);
                line += (val === 0 ? blue : white) + ch + reset + ' ';
                if (j === 2 || j === 5) line += '| ';
            }
            console.log(line);
            if (i === 2 || i === 5) console.log('------+-------+------');
        }
    }

    toJSON() {
        return this.grid;
    }

    exportTxt(filename) {
        const content = this.grid.map(row => row.join(' ')).join('\n');
        fs.writeFileSync(filename, content);
    }

    exportCsv(filename) {
        const content = this.grid.map(row => row.join(',')).join('\n');
        fs.writeFileSync(filename, content);
    }
}

program
    .option('--generate <level>', 'Generate puzzle: easy, medium, hard, expert')
    .option('--solve <file>', 'Solve puzzle from file')
    .option('--export-json <file>', 'Export to JSON')
    .option('--export-csv <file>', 'Export to CSV')
    .option('--export-txt <file>', 'Export to TXT')
    .option('--import <file>', 'Import puzzle from file')
    .option('--print', 'Print the grid')
    .parse(process.argv);

const opts = program.opts();

async function main() {
    let sudoku = null;

    if (opts.generate) {
        const s = new Sudoku();
        const solution = s.generate(opts.generate);
        sudoku = s;
        if (opts.print) {
            console.log(chalk.cyan(`Generated ${opts.generate} puzzle:`));
            s.print();
        }
        if (opts.exportJson) {
            fs.writeFileSync(opts.exportJson, JSON.stringify({ puzzle: s.grid, solution }, null, 2));
            console.log(chalk.green(`Exported to ${opts.exportJson}`));
        }
        if (opts.exportCsv) s.exportCsv(opts.exportCsv);
        if (opts.exportTxt) s.exportTxt(opts.exportTxt);
    } else if (opts.solve) {
        let data;
        if (opts.solve.endsWith('.json')) {
            data = JSON.parse(fs.readFileSync(opts.solve, 'utf8'));
            sudoku = new Sudoku(data.puzzle || data);
        } else {
            sudoku = new Sudoku();
            const content = fs.readFileSync(opts.solve, 'utf8');
            const rows = content.trim().split('\n');
            sudoku.grid = rows.map(row => row.trim().split(/\s+/).map(Number));
        }
        if (sudoku.solve()) {
            if (opts.print) {
                console.log(chalk.cyan('Solved puzzle:'));
                sudoku.print();
            }
            if (opts.exportJson) {
                fs.writeFileSync(opts.exportJson, JSON.stringify(sudoku.grid));
                console.log(chalk.green(`Exported to ${opts.exportJson}`));
            }
        } else {
            console.log(chalk.red('No solution found.'));
        }
    } else if (opts.import) {
        let data;
        if (opts.import.endsWith('.json')) {
            data = JSON.parse(fs.readFileSync(opts.import, 'utf8'));
            sudoku = new Sudoku(data.puzzle || data);
        } else {
            sudoku = new Sudoku();
            const content = fs.readFileSync(opts.import, 'utf8');
            const rows = content.trim().split('\n');
            sudoku.grid = rows.map(row => row.trim().split(/\s+/).map(Number));
        }
        if (opts.print && sudoku) {
            sudoku.print();
        }
    } else {
        program.help();
    }
}

main().catch(console.error);

File: sudoku_generator.py

#!/usr/bin/env python3
# sudoku_generator.py
import argparse
import json
import csv
import sys
import random
import math
from colorama import init, Fore, Style

init(autoreset=True)

class Sudoku:
    def __init__(self, grid=None):
        self.size = 9
        self.box_size = 3
        self.grid = grid if grid else [[0 for _ in range(9)] for _ in range(9)]

    def copy(self):
        return Sudoku([row[:] for row in self.grid])

    def is_valid(self, row, col, num):
        # Проверка строки и столбца
        for i in range(9):
            if self.grid[row][i] == num or self.grid[i][col] == num:
                return False
        # Проверка блока 3x3
        start_row, start_col = (row // 3) * 3, (col // 3) * 3
        for i in range(3):
            for j in range(3):
                if self.grid[start_row + i][start_col + j] == num:
                    return False
        return True

    def solve(self):
        for row in range(9):
            for col in range(9):
                if self.grid[row][col] == 0:
                    for num in range(1, 10):
                        if self.is_valid(row, col, num):
                            self.grid[row][col] = num
                            if self.solve():
                                return True
                            self.grid[row][col] = 0
                    return False
        return True

    def count_solutions(self, limit=2):
        count = 0
        for row in range(9):
            for col in range(9):
                if self.grid[row][col] == 0:
                    for num in range(1, 10):
                        if self.is_valid(row, col, num):
                            self.grid[row][col] = num
                            count += self.count_solutions(limit)
                            self.grid[row][col] = 0
                            if count >= limit:
                                return count
                    return count
        return 1

    def generate(self, difficulty='medium'):
        # Заполняем диагональные блоки
        self._fill_diagonal_blocks()
        self.solve()
        # Удаляем ячейки в зависимости от сложности
        cells_to_remove = {
            'easy': 30,
            'medium': 40,
            'hard': 50,
            'expert': 55
        }.get(difficulty, 40)
        
        solution = [row[:] for row in self.grid]
        positions = [(r, c) for r in range(9) for c in range(9)]
        random.shuffle(positions)
        
        removed = 0
        for r, c in positions:
            if removed >= cells_to_remove:
                break
            backup = self.grid[r][c]
            self.grid[r][c] = 0
            # Проверяем уникальность решения
            copy = self.copy()
            if copy.count_solutions(2) == 1:
                removed += 1
            else:
                self.grid[r][c] = backup
        return solution

    def _fill_diagonal_blocks(self):
        for block in range(0, 9, 3):
            numbers = list(range(1, 10))
            random.shuffle(numbers)
            idx = 0
            for i in range(block, block + 3):
                for j in range(block, block + 3):
                    self.grid[i][j] = numbers[idx]
                    idx += 1

    def is_complete(self):
        for row in self.grid:
            if 0 in row:
                return False
        return True

    def to_json(self):
        return self.grid

    @classmethod
    def from_json(cls, data):
        return cls(data)

    def print_grid(self, color=True):
        if color:
            for i, row in enumerate(self.grid):
                line = ''
                for j, val in enumerate(row):
                    if val == 0:
                        line += Fore.BLUE + '. ' + Style.RESET_ALL
                    else:
                        line += Fore.WHITE + str(val) + ' ' + Style.RESET_ALL
                    if j == 2 or j == 5:
                        line += '| '
                print(line)
                if i == 2 or i == 5:
                    print('------+-------+------')
        else:
            for i, row in enumerate(self.grid):
                line = ' '.join(str(v) if v != 0 else '.' for v in row)
                if i == 2 or i == 5:
                    print('------+-------+------')
                print(line)

    def export_txt(self, filename):
        with open(filename, 'w') as f:
            for row in self.grid:
                f.write(' '.join(str(v) if v != 0 else '0' for v in row) + '\n')

    def export_csv(self, filename):
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            for row in self.grid:
                writer.writerow(row)

def main():
    parser = argparse.ArgumentParser(description="Sudoku Generator")
    parser.add_argument("--generate", choices=['easy', 'medium', 'hard', 'expert'], help="Generate puzzle")
    parser.add_argument("--solve", help="Solve puzzle from file")
    parser.add_argument("--export-json", help="Export to JSON")
    parser.add_argument("--export-csv", help="Export to CSV")
    parser.add_argument("--export-txt", help="Export to TXT")
    parser.add_argument("--import", dest="import_file", help="Import puzzle from file")
    parser.add_argument("--print", action="store_true", help="Print the grid")
    args = parser.parse_args()

    sudoku = None

    if args.generate:
        s = Sudoku()
        solution = s.generate(args.generate)
        sudoku = s
        if args.print:
            print(f"Generated {args.generate} puzzle:")
            sudoku.print_grid()
        if args.export_json:
            with open(args.export_json, 'w') as f:
                json.dump({"puzzle": sudoku.grid, "solution": solution}, f)
            print(f"Exported to {args.export_json}")
        if args.export_csv:
            sudoku.export_csv(args.export_csv)
        if args.export_txt:
            sudoku.export_txt(args.export_txt)

    elif args.solve:
        # Загрузка из файла
        if args.solve.endswith('.json'):
            with open(args.solve, 'r') as f:
                data = json.load(f)
                grid = data.get('puzzle', data)
                sudoku = Sudoku(grid)
        else:
            sudoku = Sudoku.from_txt(args.solve)
        if sudoku.solve():
            if args.print:
                print("Solved puzzle:")
                sudoku.print_grid()
            if args.export_json:
                with open(args.export_json, 'w') as f:
                    json.dump(sudoku.grid, f)
                print(f"Exported to {args.export_json}")
        else:
            print("No solution found.")

    elif args.import_file:
        if args.import_file.endswith('.json'):
            with open(args.import_file, 'r') as f:
                data = json.load(f)
                grid = data.get('puzzle', data)
                sudoku = Sudoku(grid)
        else:
            sudoku = Sudoku.from_txt(args.import_file)
        if args.print and sudoku:
            sudoku.print_grid()

    else:
        parser.print_help()

if __name__ == "__main__":
    main()
